Support NUMA

yangzhon opened the pull request:
@rbradford @sameo @mcastelino 

This patchset enable NUMA function in the acpi table and we can see some numa info from guest kernel log
[    0.000000] ACPI: SRAT 0x000000001FFC1038 000150 (v01 BOCHS  BXPCSRAT 00000001 BXPC 00000001)
[    0.000000] ACPI: MCFG 0x000000001FFC1188 00003C (v01 BOCHS  BXPCMCFG 00000001 BXPC 00000001)
[    0.000000] ACPI: Local APIC address 0xfee00000
[    0.000000] SRAT: PXM 0 -> APIC 0x00 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x01 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x02 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x03 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x04 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x05 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x06 -> Node 0
[    0.000000] SRAT: PXM 0 -> APIC 0x07 -> Node 0
[    0.000000] ACPI: SRAT: Node 0 PXM 0 [mem 0x00000000-0x0009ffff]
[    0.000000] ACPI: SRAT: Node 0 PXM 0 [mem 0x00100000-0x1fffffff]
[    0.000000] ACPI: SRAT: Node 0 PXM 0 [mem 0x100000000-0x19fffffff] hotplug
[    0.000000] NUMA: Node 0 [mem 0x00000000-0x0009ffff] + [mem 0x00100000-0x1ffbffff] -> [mem 0x00000000-0x1ffbffff]

Please help review those patches, thanks!

fixed  #35

yangzhon committed:
virt: acpi: move build_srat() to aml-build.c

Move build_srat() related function into aml-build.c for
different platform.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

rbradford commented on hw/acpi/aml-build.c:
```diff
@@ -2388,9 +2388,7 @@ void build_srat(GArray *table_data, BIOSLinker *linker,
-    ram_addr_t hotplugabble_address_space_size =
-        object_property_get_int(OBJECT(machine), PC_MACHINE_MEMHP_REGION_SIZE,
-                                NULL);
+    ram_addr_t hotplugabble_address_space_size = conf->hotplug_memory_size;
```
Is it correct to assign a uint32 to a ram_addr_t ?

rbradford commented on hw/i386/pc.c:
```diff
@@ -1804,10 +1795,6 @@ static void pc_machine_class_init(ObjectClass *oc, void *data)
     nc->nmi_monitor_handler = x86_nmi;
     mc->default_cpu_type = TARGET_DEFAULT_CPU_TYPE;
 
-    object_class_property_add(oc, PC_MACHINE_MEMHP_REGION_SIZE, "int",
```
You can't do this as the object properties are exposed on the command line as QEMU "API". Doing this breaks users of the pc machine type. Please find a different way to solve your problem.

rbradford commented on include/hw/i386/virt.h:
```diff
@@ -62,6 +62,7 @@ typedef struct {
     HotplugHandler *acpi_dev;
 
     MemoryHotplugState hotplug_memory;
+    uint32_t hotplug_memory_size;
```
It annoys me that we duplicate another variable between VirtMachineState and AcpiConfiguration. the VirtMachineState struct has a pointer to the AcpiConfiguration so it can always access that instead. Maintaining state between the AcpiConfiguration struct and the VirtMachineState struct is a source of bugs.

yangzhon committed:
virt: acpi: add numa support

Build srat table to send the numa topology to the BIOS, which has to
setup the tables. The guest kernel can normally show numa acpi info.
Since the default numa code is still here, when we build srat table for
numa, we can adds a -numa command line parameter in virt platform and
expose those numa nodes to vm.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon commented:
@rbradford 

I addressed your comments and re-sent the PR, please review again.  This new PR also addressed some conflict in semaphore. thanks!

rbradford commented:
@yangzhon The only feedback I have is i'd like more explanation in the commit message for the final commit about how you enable it (new tables) and what benefit it brings.

yangzhon commented:
@rbradford,

Changed the commit info, thanks!
